### DISCUSSION TITLE
Get action view contents only without rendering the view on browser when using dispatch()

### DISCUSSION THREAD

**Discussion by Lumu-web** [phalcon/cphalcon#15427] [Q&A]
I have the following code below which works as I suppose it should. I would like to know if there is a way to just return the view contents without actually showing the view on the browser. At the moment, the dispatcher runs and targets the controller/action which has a view associated with that action. At the end of the dispatch, that view echoes in the browser.

I would like to be able to run this method multiple times with different contoller/actions and just collect the view contents in order to finally send them to another view where they will be combined.

Collecting the contents succeed but the ultimate view cant be displayed as the page ends up showing the view of the last action the dispatch hit. Any help in stopping that view from actually loading?

getData(){
$dispatcher = $application->dispatcher;
    $dispatcher->setNamespaceName(xxxxx);
    $dispatcher->setModuleName(xxxxx);
    $dispatcher->setControllerName(xxxxx);
    $dispatcher->setActionName(xxxxx);
    /**
     * @var \Phalcon\Mvc\ViewInterface $view
     */
    $view = $dispatcher->dispatch()->view;

    $view->start();

    $view->render(
        $dispatcher->getControllerName(),
        $dispatcher->getActionName(),
        $dispatcher->getParams()
    );

    $view->finish();

    return $view->getContent();
    }

**Comment by quasipickle** [comment]
Off the top of my head, I suppose in your last action you could disable the view, then just `echo()` the captured content.

However, I think you're mis-using how controllers & actions are supposed to work. Only 1 action should be fully executed per request - the whole system is set up that way.

If you're finding the output (and consequently the logic) from an action is needed in multiple places, you should move the logic into a component.

**Reply by Lumu-web** [reply]
Thanks,  view->disable() disables the view but does not display the final view once all contents are gathered. . So I am left with a blank screen. I guess the flow never returns back to the initial action that started it all. 

Yea unfortunately I found the system already coded that way. And now that updates are needed, recoding seems like a lot more work so we basically wanted a work around to what was already there. thank you very much for your reply. 

Might just consider a recoding if no work arounds exist

**Reply by quasipickle** [reply]
Well, I don't think there's an easy fix.  Presumably you also want these actions to also be executable themselves?

The only thing that comes to mind is you add a `captureOutput` flag to your base controller.  Then, in each of these actions, if `captureOutput` is true, disable the view, set up an output buffer, call the necessary view manually, then store the output buffer in a session.  Then in your last action, gather up all the output stored in the session and output it.

Alternatively, you could write your own view layer that intercepts call to render() and does that logic itself.

I'm not aware of any built-in flag that lets you toggle the default behaviour between outputting and capturing.

**Comment by borisdelev** [comment]
hmm... i found old code of mine and i accomplish that way:

```php
        $view = clone $this->view;
        $view->start();
        $view->setLayout('emails');
        $view->setVars($params);
        $view->render('_emails', $template);
        $view->finish();
        $message = $view->getContent();
```

**Comment by Jeckerson** [comment]
You can try to use Simple View. With it, it is possible to pick template, pass variables and render into variable or output to the browser.

https://docs.phalcon.io/4.0/en/views#simple-rendering
